### PR TITLE
Updated `assert_character_scalar()` & `assert_s3_class()` with `cli::cli_abort()` messaging

### LINKED ISSUE #375: `assert_character_scalar()` (depends on `assert_character_vector()` and `assert_logical_scalar()`)

### LINKED ISSUE #377: `assert_s3_class()` (depends on `assert_logical_scalar()` and `assert_character_scalar()`)

### PR DESCRIPTION
**Pull request by ddsjoberg**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/dev/CONTRIBUTING.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #375 and closes #377 into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiraldev/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by millerg23** [comment]
@ddsjoberg looks good, only comment I have is for `assert_character_scalar` we use the `.val` option which users separator "and" for last 2 values returned, can this be replaced with "or"?? 
```
Argument `msg_type` must be equal to one of "warning" and "error".
```

I tried replacing `.val` with `.or` but then the values aren't quoted!
```
Argument `msg_type` must be equal to one of warning or error.
```

**Comment by ddsjoberg** [comment]
Thank you @millerg23 for reviewing!!

I agree, "or" would be preferable when the values are properly quoted. Here are a few options:
1. Stick with it as it is, since the message is clear enough and we get nice quotations and colors.
2. Use `{.or}` so we get "or" instead of "and" in the list. But we won't get quotations or coloring.
3. Use `rlang::arg_match0()` function that does both quoting and coloring (and even adds on improved messaging including what the user passed that is not valid). But we can't modify the error message.

```r
check_valid_fruit <- function(fruit) {
    rlang::arg_match0(fruit, c("apples", "bananas"))
}

check_valid_fruit("apples")
#> [1] "apples"
check_valid_fruit("lemons")
#> Error in `check_valid_fruit()`:
#> ! `fruit` must be one of "apples" or "bananas", not "lemons".
```

4. The last option is to change the default behavior of `{.val}` by creating a div container and defining different behavior. Admittedly, I've steered away from this because I friend told me they did this then {cli} made a breaking change to the implementation and it broke their code. But it is the solution that gets us exactly what we are looking for.

``` r
# define a div and change the .val instructions.
divid <- cli::cli_div(theme = list(.val = list("vec-last" = ", or ")))
cli::cli_text("{.val {letters[1:3]}}")
#> "a", "b", or "c"
cli::cli_end(divid)

# testing that the div is closed after `testing()` function stops execution
testing <- function() {
  divid <- cli::cli_div(theme = list(.val = list("vec-last" = ", or ")))
  cli::cli_abort("{.val {letters[1:3]}}") 
}
testing()
#> Error in `testing()`:
#> ! "a", "b", or "c"

# reverts back to default "and" messaging
cli::cli_text("{.val {letters[1:3]}}")
#> "a", "b", and "c"
```



I am fine with _any_ of these solutions as each of them provide clear messaging to the user. What do you all think?

**Comment by millerg23** [comment]
@ddsjoberg thanks for providing the options, I think I prefer option 3 for `assert_character_scalar`. Would discount option 2, my initial thoughts was along the lines of option 4, but as you say may cause us issues down the line.
So for me in order of preference 3 > 4 > 1 > 2, but won't lose sleep over the option we choose ;-)

**Comment by bms63** [comment]
Could we hold off and discuss on Wednesday! I really want to get more involved, but the study grind is punishing right now :(.

**Comment by ddsjoberg** [comment]
> Could we hold off and discuss on Wednesday! I really want to get more involved, but the study grind is punishing right now :(.

Sure @bms63 !

**Comment by manciniedoardo** [comment]
> @ddsjoberg thanks for providing the options, I think I prefer option 3 for `assert_character_scalar`. Would discount option 2, my initial thoughts was along the lines of option 4, but as you say may cause us issues down the line. So for me in order of preference 3 > 4 > 1 > 2, but won't lose sleep over the option we choose ;-)

3 >>> 1 > 4 > 2 is my preferred order. I think that retaining the "standard" error message is ok, especially for a simple case like `assert_character_scalar()`. I would veer away from something more complicated like @ddsjoberg described in 4 as this may involve more maintenance for not much practical gain.

File: R/assertions.R [hunk 1]
```diff
@@ -132,30 +133,41 @@ assert_data_frame <- function(arg,
 assert_character_scalar <- function(arg,
                                     values = NULL,
                                     case_sensitive = TRUE,
-                                    optional = FALSE) {
+                                    optional = FALSE,
+                                    arg_name = rlang::caller_arg(arg),
+                                    message = NULL,
```
Comment (bundfussr) [inline R/assertions.R:138]: I think we need to rethink the `message` argument. It works well for assertions which issue a single message only. But if different messages are issued like "the argument is of wrong type" and "the argument is of correct type but the values are not as expected", it does not work.

Maybe we should remove the `message` and `class` argument. And if it it necessary to change the message or class, the error should be caught and modified.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/assertions.R [hunk 2]
```diff
@@ -132,30 +133,41 @@ assert_data_frame <- function(arg,
...
   }
 
+  message <-
+    message %||%
+    ifelse(
+      is.null(values),
+      "Argument {.arg {arg_name}} must be a scalar of class {.cls character},
+       but is {.obj_type_friendly {arg}}.",
+      "Argument {.arg {arg_name}} must be equal to one of {.val {values}}."
+    )
```
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/assertions.R:155]: I agree we should absolutely talk about the message argument, thanks @bundfussr !

If we were designing these checks from scratch, I would have advocated for simpler checks rather than compound checks. But now that we have compound checks, I am in favor of compound default messages still with the ability to modify the message via an argument. As more admiral-flavored packages are built, they may need to change the messages in certain situations and the message argument gives them that flexibility.

In the example above, I didn't add the class to the message because I thought that by listing the values that are accepted it conveys the information the user needs. If we feel we that we should list the values and the class of these values, that can be done. But I personally don't think that is needed to message to the users how they should update their code.

File: R/assertions.R [hunk 3]
```diff
@@ -132,30 +133,41 @@ assert_data_frame <- function(arg,
...
     return(invisible(arg))
   }
 
+  message <-
+    message %||%
+    ifelse(
+      is.null(values),
+      "Argument {.arg {arg_name}} must be a scalar of class {.cls character},
+       but is {.obj_type_friendly {arg}}.",
+      "Argument {.arg {arg_name}} must be equal to one of {.val {values}}."
```
Comment (kaz462) [inline R/assertions.R:154]: should "one of" be removed when `values` contains one value
```r
> assert_character_scalar("1", "a")
Error:
! Argument `"1"` must be equal to one of "a".
Run `rlang::last_error()` to see where the error occurred.
```

File: R/assertions.R [hunk 4]
```diff
@@ -132,30 +133,41 @@ assert_data_frame <- function(arg,
...
   assert_logical_scalar(optional)
 
   if (optional && is.null(arg)) {
     return(invisible(arg))
   }
 
+  message <-
+    message %||%
+    ifelse(
+      is.null(values),
```
[the remark this answers is not in the document]
Comment (millerg23) [reply R/assertions.R:151]: Do we need `is.null(values)` ? Could it be replaced with `!is.character(arg) | length(arg) != 1L`.
So check argument is of correct type, if so, then check it is in allowed list of values.

[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/assertions.R:151]: thanks for the review comments. 

I updated the two default error messages to have identical information, and when `values` are passed the message is supplemented with the additional information for which values are allowed. 

```r
ifelse(
      is.null(values),
      "Argument {.arg {arg_name}} must be a scalar of class {.cls character},
       but is {.obj_type_friendly {arg}}.",
      "Argument {.arg {arg_name}} must be a scalar of class {.cls character} and
       equal to one of {.val {values}}."
    )
```

@millerg23 I updated the chunk with the or syntax you recommended.

```r
 if (!is.character(arg) || length(arg) != 1L) {
    cli::cli_abort(
      message = message,
      call = call,
      class = c(class, "assert-admiraldev")
    )
  }
```

File: R/assertions.R (hunk 3 above)
Comment (ddsjoberg) [reply R/assertions.R:154]: I don't think this is the type of check a developer would write to check for a single value? Is there an example of this in admiral?

File: R/assertions.R [hunk 5]
```diff
@@ -132,30 +133,41 @@ assert_data_frame <- function(arg,
...
   if (length(arg) != 1L) {
-    err_msg <- sprintf(
-      "`%s` must be a character scalar but is a character vector of length %d",
-      arg_name(substitute(arg)),
-      length(arg)
+    cli::cli_abort(
+      message = message,
+      call = call,
+      class = c(class, "assert-admiraldev")
     )
```
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/assertions.R:170]: Good point, I combined the two calls to `cli_abort()` into one. Thanks

File: R/assertions.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/assertions.R:155]: I updated the message to re-iterate that the value must be scalar character.

```r
ifelse(
      is.null(values),
      "Argument {.arg {arg_name}} must be a scalar of class {.cls character},
       but is {.obj_type_friendly {arg}}.",
      "Argument {.arg {arg_name}} must be a scalar of class {.cls character} and
       equal to one of {.val {values}}."
    )
```

File: R/assertions.R (hunk 1 above)
Comment (ddsjoberg) [reply R/assertions.R:138]: After meeting to discuss, we will keep the `message` argument. Thanks for the good discussion point!

**Comment by ddsjoberg** [comment]
> 4. The last option is to change the default behavior of {.val} by creating a div container and defining different behavior. Admittedly, I've steered away from this because I friend told me they did this then {cli} made a breaking change to the implementation and it broke their code. But it is the solution that gets us exactly what we are looking for.

FYI I asked my friend about this and they were doing two things at once. Building new cli commands like this `{.xxx arg}` and changing the divid. The breaking change they experienced was from defining new cli commands, not the divid change. Also, this was a number of years ago already. With that in mind, I updated the error message to use the modified `cli_div()` to get the exact message we want.

File: R/assertions.R (hunk 3 above)
Comment (kaz462) [reply R/assertions.R]: You are right - do you think it's necessary to add a length restriction >1 for the input `values`?

Comment (ddsjoberg) [reply R/assertions.R]: Hmmm good question. I think we can leave it as it is so we don't make unnecessary changes to the current implementation of the check, rather focussing on the messaging to the users.

If we want to make this change, perhaps we can make that change in another PR that focusses on that change to the functionality?

**Review by kaz462 — APPROVED** [review]

**Comment by bundfussr** [comment]
Why has the content of the error messages been changed?
Before we had:
```
> assert_character_scalar(arg <- c("first", "last"), values = c("first", "last"))
Error in `assert_character_scalar()`:
! `arg` must be a character scalar but is a character vector of length 2
Run `rlang::last_trace()` to see where the error occurred.
> assert_character_scalar(arg <- 1, values = c("first", "last"))
Error in `assert_character_scalar()`:
! `arg` must be a character scalar but is `1`
Run `rlang::last_trace()` to see where the error occurred.
> assert_character_scalar(arg <- "1", values = c("first", "last"))
Error in `assert_character_scalar()`:
! `arg` must be one of 'first' or 'last' but is '1'
Run `rlang::last_trace()` to see where the error occurred.
```
And now we have:
```
> assert_character_scalar(arg <- c("first", "last"), values = c("first", "last"))
Error:
! Argument `arg <- c("first", "last")` must be a scalar of class <character> and equal to one of
  "first" or "last".
Run `rlang::last_trace()` to see where the error occurred.
> assert_character_scalar(arg <- 1, values = c("first", "last"))
Error:
! Argument `arg <- 1` must be a scalar of class <character> and equal to one of "first" or "last".
Run `rlang::last_trace()` to see where the error occurred.
> assert_character_scalar(arg <- "1", values = c("first", "last"))
Error:
! Argument `arg <- "1"` must be a scalar of class <character> and equal to one of "first" or
  "last".
Run `rlang::last_trace()` to see where the error occurred.
```
The new messages are longer but contain less information. The message is always the same regardless of what caused the failure.
I would consider the old messages more user friendly.
Could we use `cli_abort()` but keep the old content of the messages?
